crates/jit/src/instantiate.rs

```diff
@@ -399,43 +402,102 @@ impl CompiledModule {
     /// about new code that is loaded.
     pub fn from_artifacts(
         mmap: MmapVec,
-        info: Option<CompiledModuleInfo>,
+        mut info: Option<CompiledModuleInfo>,
         profiler: &dyn ProfilingAgent,
         id_allocator: &CompiledModuleIdAllocator,
     ) -> Result<Self> {
+        use gimli::SectionId::*;
+
+        // Parse the `code_memory` as an object file and extract information
+        // about where all of its sections are located, stored into the
+        // `CompiledModule` created here.
+        //
+        // Note that dwarf sections here specifically are those that are carried
+        // over directly from the original wasm module's dwarf sections, not the
+        // wasmtime-generated host DWARF sections.
         let obj = File::parse(&mmap[..]).context("failed to parse internal elf file")?;
-        let opt_section = |name: &str| obj.section_by_name(name).and_then(|s| s.data().ok());
-        let section = |name: &str| {
-            opt_section(name)
-                .ok_or_else(|| anyhow!("missing section `{}` in compilation artifacts", name))
-        };
+        let mut wasm_data = None;
+        let mut address_map_data = None;
+        let mut func_name_data = None;
+        let mut trap_data = None;
+        let mut code = None;
+        let mut dwarf_sections = Vec::new();
+        for section in obj.sections() {
+            let name = section.name()?;
+            let data = section.data()?;
+            let range = subslice_range(data, &mmap);
+            let mut gimli = |id: gimli::SectionId| {
+                let idx = id as usize;
+                if dwarf_sections.len() <= idx {
+                    dwarf_sections.resize(idx + 1, 0..0);
+                }
```

Comment:
I kinda wish you could pre-allocate the `dwarf_sections` for the maximum possible `SectionId`, since there aren't that many of them, rather than having to think about the edge cases of incremental resizing. But I don't see any easy way to get the max `SectionId` value.

I think this condition would read more clearly to me if flipped (`if idx >= dwarf_sections.len()`). But I always have to think about this kind of condition either way, so I don't feel strongly about it.

Reply:
Ah yeah in terms of minor inefficiencies in loading a precompiled artifact I think this is pretty low down on the list, all the other operations I think vastly dominate the time spent doing that.

Reply:
Oh, absolutely. I wasn't worried about inefficiency of resizing the vector. This `if` condition was the only bit I had to stop to think about, and also the only place where you had a bug caught by CI. So it would be nice if we could pre-size the array to avoid having to think about the correctness of this `if` statement. But there doesn't appear to be a good way to do that, and after all CI _did_ catch the bug, so the way you have it is probably the best option.